### PR TITLE
Improved support for custom image loaders

### PR DESCRIPTION
**Pull request by supagu**
Removed image loader limit. Support to add/remove/find/list image loaders. Resolves both #19322 and #19037

### CODE REVIEW

**Comment by supagu** [comment]
Sample usage in my heightmap loading class:

`Ref<StreamTexture> n;
    if (path.ends_with("png")) {
        ImageFormatLoader* orig_loader = ImageLoader::recognize("png");
        ImageLoader::remove_image_format_loader(orig_loader);

        BImageLoaderPNG new_loader;
        ImageLoader::add_image_format_loader(&new_loader);

        n = ResourceLoader::load(path);

        ImageLoader::remove_image_format_loader(&new_loader);
        ImageLoader::add_image_format_loader(orig_loader);
    }
    else {
        n = ResourceLoader::load(path);
    }`

**Comment by supagu** [comment]
Ops, I've also added in 16-bit PNG support which looks to have been caught up in this PR.
I would like both to go in but will separate if required.

This will contribute to help #17450, which is what I am doing: loading my heightmap as a single channel 16-bit PNG (which is faster and I suspect, also more memory efficient than EXR as it will come in as a single channel).

**Comment by mhilbrunner** [comment]
You should remove the second commit and create a second PR with it :)

**Comment by supagu** [comment]
Fixed

File: core/io/image_loader.cpp [hunk 1]
```diff
@@ -82,28 +82,36 @@ Error ImageLoader::load_image(String p_file, Ref<Image> p_image, FileAccess *p_c
...
+			return loader[i];
 	}
 
-	return false;
+	return NULL;
 }
 
-ImageFormatLoader *ImageLoader::loader[MAX_LOADERS];
-int ImageLoader::loader_count = 0;
+Vector<ImageFormatLoader *> ImageLoader::loader;
```
Comment (reduz) [inline core/io/image_loader.cpp]: This vector needs to be cleared at some point at exit, or its memory will be leaked (or Godot crash)

**Review by reduz — CHANGES_REQUESTED** [review]

File: core/io/image_loader.cpp (hunk 1 above)
Comment (supagu) [reply core/io/image_loader.cpp]: @reduz 
Would you prefer having something like ImageLoader::cleanup() which does this (if so, where should it go?) 
OR
make it the responsibility of the calling code that calls ImageLoader::add_image_format_loader
to remove its image loader before it destroys the image loader
(I think this second option is safer)

Comment (supagu) [reply core/io/image_loader.cpp]: @reduz, this is now fixed - I've added calls to remove the image loader which will free the vector once it resizes itself to zero

**Comment by supagu** [comment]
Fixed by adding remove image loader calls which will free the vector once it reaches zero

**Comment by supagu** [comment]
is there anything I need to do here to say I've made further changes as requested and again ready for review?

**Comment by reduz** [comment]
Sorry, I missed this
I think a cleanup function would probably be better, to avoid  have to remember removing the loaders

**Comment by supagu** [comment]
I've attempted to make the appropriate changes, but there is issues now with merging. I'm not sure how to fix this now as Merge into master didn't work/compile....

**Comment by supagu** [comment]
I've blown this away and made a new PR #21716

**Pull request closed without merging**